### PR TITLE
[WIP] - Support of a subset of part 8 "DataAccess"

### PR DESCRIPTION
**Pull request by Nik625**
With the content of this pull requests a subset of the OPC UA Specification Part 8 "DataAccess" is supported. The subset is defined among the following Conformance Units (as defined in the Compliance Test Tool): _AnalogItemType_, _AnalogArrayType_, _DataItemType_, _DiscreteItemType_, _SemanticsChanged_ and _PercentDeadbandFilter_. The "_ValuePrecision_" for variables of type "_DateTime_" is only implemented as a very first draft.

See also #2267 for the feature branch.

### CODE REVIEW

File: arch/ua_network_tcp.c [hunk 1]
```diff
@@ -525,20 +525,22 @@ UA_ServerNetworkLayer
 UA_ServerNetworkLayerTCP(UA_ConnectionConfig config, UA_UInt16 port, UA_Logger logger) {
     UA_ServerNetworkLayer nl;
     memset(&nl, 0, sizeof(UA_ServerNetworkLayer));
+    nl.deleteMembers = ServerNetworkLayerTCP_deleteMembers;
```
Comment (basyskom-meerkoetter) [inline arch/ua_network_tcp.c]: does this "belong" into this series or is this something to break out into a PR of its own?
This is a fix to an overall(?) issues found by oss-fuzz?

File: examples/tutorial_server_dataaccess.c [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+/**
+ * Integrating a DataAccess example VarieableType with Properties
+ * DataAccess defines new Variable Types and Properties for describing real time data.
+ * In this example a VariableType node is added with its respective DataAccess Properties
+*/
+#include "open62541.h"
+#include <signal.h>
+#include "ua_server_dataaccess.h"
+
+UA_Boolean running = true;
```
Comment (basyskom-meerkoetter) [inline examples/tutorial_server_dataaccess.c]: static UA_Boolean running = true
to limit the scope of the symbol to this translation unit (just like it is already done for the functions)?

File: examples/tutorial_server_dataaccess.c [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+int
+main(void) {
+    signal(SIGINT,  stopHandler);
+    signal(SIGTERM, stopHandler);
+
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    UA_ServerConfig *config = UA_ServerConfig_new_default();
+    UA_Server *server = UA_Server_new(config);
+
+    retval |= addDataAccessNode(server);
```
Comment (basyskom-meerkoetter) [inline examples/tutorial_server_dataaccess.c]: does it make sense to enter the server loop if addDataAccessNode() failed?

File: examples/tutorial_server_dataaccess.c [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    retval |= UA_Server_run(server, &running);
+
+    UA_Server_delete(server);
+    UA_ServerConfig_delete(config);
+    return (int)retval;
+
```
Comment (basyskom-meerkoetter) [inline examples/tutorial_server_dataaccess.c]: superfluous empty line (before })

File: src/server/ua_services_attribute.c [hunk 5]
```diff
@@ -1050,6 +1058,270 @@ writeValueAttributeWithRange(UA_VariableNode *node, const UA_DataValue *value,
...
+                const UA_VariableNode* fNode = (const UA_VariableNode*) UA_Nodestore_get(server, &node->nodeId);
+                newNotification->data.value.hasValue = true;
+                newNotification->mon = mon;
+                UA_DataValue_copy(&(fNode->value.data.value), &newNotification->data.value);
+                newNotification->data.value.hasStatus = true;
+                newNotification->data.value.status = UA_STATUSCODE_GOOD;
+                UA_Nodestore_release(server, (const UA_Node*)fNode);
+                /* Enqueue the new notification */
+                UA_Notification_enqueue(server, sub, mon, newNotification);
+			    return UA_STATUSCODE_GOOD;
```
Comment (basyskom-meerkoetter) [inline src/server/ua_services_attribute.c]: return is indented wronly

File: arch/ua_network_tcp.c (hunk 1 above)
Comment (basyskom-meerkoetter) [inline arch/ua_network_tcp.c]: does this "belong" into this series or is this something to break out into a PR of its own?
This is a fix to an overall(?) issues found by oss-fuzz?

File: examples/tutorial_server_dataaccess.c (hunk 2 above)
Comment (basyskom-meerkoetter) [inline examples/tutorial_server_dataaccess.c]: static UA_Boolean running = true
to limit the scope of the symbol to this translation unit (just like it is already done for the functions)?

File: examples/tutorial_server_dataaccess.c [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+main(void) {
+    signal(SIGINT,  stopHandler);
+    signal(SIGTERM, stopHandler);
+
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    UA_ServerConfig *config = UA_ServerConfig_new_default();
+    UA_Server *server = UA_Server_new(config);
+
+    retval |= addDataAccessNode(server);
+    retval |= UA_Server_run(server, &running);
```
Comment (basyskom-meerkoetter) [inline examples/tutorial_server_dataaccess.c]: does it make sense to enter the server loop if addDataAccessNode() failed?

File: examples/tutorial_server_dataaccess.c (hunk 4 above)
Comment (basyskom-meerkoetter) [inline examples/tutorial_server_dataaccess.c]: superfluous empty line (before })

File: src/server/ua_services_attribute.c [hunk 7]
```diff
@@ -1050,6 +1058,270 @@ writeValueAttributeWithRange(UA_VariableNode *node, const UA_DataValue *value,
...
+    if(UA_STATUSCODE_GOOD == bpr.statusCode || bpr.targetsSize >= 1) {
+        UA_Variant tVar;
+        UA_Variant_init(&tVar);
+        retval = UA_Server_readValue(server, bpr.targets->targetId.nodeId, &tVar);
+        if(retval == UA_STATUSCODE_GOOD) {
+            UA_Range* euRange = (UA_Range*) tVar.data;
+            retval = UA_Variant_checkRange(&(value->value), euRange);
+            if(retval == UA_STATUSCODE_GOOD) {
+                writeValueAttributeWithoutRange(node, &adjustedValue);
+                enqueueSemanticChangedNotification(server, node, session, UA_STATUSCODE_GOOD| 0x4000);
```
Comment (basyskom-meerkoetter) [inline src/server/ua_services_attribute.c]: what is | 0x4000?

**Comment by jpfr** [comment]
Hello there,

can you indicate when another round of reviews is required?

**Comment by Pro** [comment]
When #2267 is merged, this should be rebased onto master

**Comment by Nik625** [comment]
Unfortunately due to a lot of work, this will be continued approximately in mid of february from my side.

**Comment by Nik625** [comment]
Regarding the fragmentation into several feature PRs i only see the **SemanticsChanged** feature existing alongside the DataAccess extension. Part 8 of the specification  references this as _"Servers that implement Data Access shall set this Bit in notifications if certain Properties defined in this standard change."_ and the changed properties where the bit is set are DataAccess specific properties. Thus, I see this feature integrated in the general DataAccess functionalities.

**Comment by Nik625** [comment]
@jpfr Maybe you can explain again what you meant with splitting up Core DA functionalities and the information model. Many core functionalities rely on the existence of the VariableTypes defined in the informationmodel. Thanks in advance!

**Comment by jpfr** [comment]
`UA_ENABLE_DA` is a very coarse-grained build flag.
It could be that some people want deadband filtering. But **not** the full gamut of DA features.

This could be `UA_ENABLE_SUBSCRIPTIONS_DEADBAND` or similar.

Then we can have a list of individual features that collectively cover the DA specification.
And maybe a meta-build-flag `UA_ENABLE_DA` that enables all of them.

Can you make a list of such feature-flags?
Then we can merge the features individually.

I would like to avoid merging a 5k line monster PR in one go.

**Comment by Nik625** [comment]
image


The build flags are defined as follows:

- **UA_ENABLE_DA** remains as the meta-build flag for the entire DA functionality (this requires _UA_ENABLE_DA_INFORMATIONMODEL_ to be enabled to depict the demo nodes in the address space)
- **UA_ENABLE_DA_INFORMATIONMODEL** enables the xml-descriptions of the types and sets the informationmodel. No other enabled build flags are compulsory for this functionality to work.
- **UA_ENABLE_DA_ANALOGITEMS** enables only the AnalogItemType with its underlying properties but also sets the required informationmodel. The datatypes are referenced by a separate file datatypes_da_analog.txt. This functionality requires the informationmodel to depict the demo nodes in the address space. If no nodes are created in the adress space, this functionality would still be available but would obviously have no use. This is why it should not be activated without enabling - _UA_ENABLE_DA_INFORMATIONMODEL_.

- **UA_ENABLE_DA_SEMANTICCHANGED** enables the enqueuing of new data change notifications with their semanticchanged bit set in the statuscode. This function requires the dataypes to be referenced by the respective file (datatypes_da_analog.txt.). In its core functionality the implementation checks for the existence of relevant underlying properties of the AnalogItemType which are modelled by the informationmodel (such as _Range_). Thus _UA_ENABLE_DA_INFORMATIONMODEL_ needs to be switched on.
- **UA_ENABLE_SUBSCRIPTIONS_DEADBAND** enables the percent deadband data change filter for MonitoredItems in Subscriptions. The percentage is relative to the defined range property and thus requires the range datatype to be enabled through setting the datatype file (e.g _UA_ENABLE_ANALOGITEMS_). The user needs to create this property as a variable node and thus needs the xml definition of the according range datatype (Enable of _UA_ENABLE_DA_INFORMATIONMODEL_ is required in this case aswell).
- The _greyed out build flags_ are shown to fulfill the depiction. Their mere representation is again coppled to the informationmodel, but besides these flags do not define new datatypes that could exist independent of the other parts. Thus, integrating such build flags is not recommended.

The given structure makes it difficult to find appropriate, independent build flags to separate the functionality. Separating the different datatypes would still result in the need of integrating a partial informationmodel in order to be able to actually create these datatypes and use the functionalities behind them.

Ultimately, a separation into independent features could be realized internally, but users would need to enable more than their desired build flag to benefit from the functionalities.

**Comment by jpfr** [comment]
@Nik625 Great to see this pick up again!

The PR is marked as [Review]. There are now a lot of "fix commits".
Can you squash/rebase them out to have a clean history?

The dependency graph you show is really nice. Because it is a DAG (directed acyclic graph).
So we can have features that "build on top of each other".
It should be possible to have one commit per feature in this PR:

- Rebase on the current master
- Squash all commits into a single commit.
- Do a `git reset --soft HEAD^`, then a `git reset` to get all changes as "unadded/unstaged changes"
- Add selectively with `git add -p` and commit

**Comment by ccvca** [comment]
@Nik625 Do you closed this by accident?

**Pull request closed without merging**